Bug: flat config with "eslint:all" on a .cjs file recommends function form of "use strict" instead of global form

### Environment

Node version: `v16.14.0`
npm version: `8.3.1`
Local ESLint version: `v8.23.1`
Global ESLint version: `zsh: command not found: eslint`
Operating System: `Fedora release 36 (Thirty Six)`


### What parser are you using?

Default (Espree)

### What did you do?

## package.json

Notice `type` is `module`.

```js
{
  "devDependencies": {
    "eslint": "^8.23.1"
  },
  "license": "ISC",
  "name": "eslint-all-cjs",
  "scripts": {
    "lint": "eslint --fix ."
  },
  "type": "module",
  "version": "1.0.0"
}
```

## eslint.config.js

Using the new flat config file format.  The default `languageOptions` `sourceType` should be `commonjs` for `.cjs` files.

```js
export default ["eslint:all"];
```

## index.cjs

A `.cjs` file that ESLint should consider to be a `commonjs` file.

```js
module.exports = function identity (value) {

    return value;

};
```



### What did you expect to happen?

Expected: a message to use the global form of `"use strict"`.

Why: the docs for rule `strict` say:
> This rule has a string option:
> * "safe" (default) corresponds either of the following options:
>   * "global" if ESLint considers a file to be a CommonJS module
>   * "function" otherwise

The file ends with `.cjs`, so ESLint should recommend "global".

### What actually happened?

A message to use the function form:
```
/path/to/eslint-all-cjs/index.cjs
  1:18  error  Use the function form of 'use strict'  strict

✖ 1 problem (1 error, 0 warnings)
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

--- a/lib/rules/strict.js
+++ b/lib/rules/strict.js
@@ -105,7 +105,7 @@ module.exports = {
         if (ecmaFeatures.impliedStrict) {
             mode = "implied";
         } else if (mode === "safe") {
-            mode = ecmaFeatures.globalReturn ? "global" : "function";
+            mode = ecmaFeatures.globalReturn || context.languageOptions.sourceType === "commonjs" ? "global" : "function";
         }
 
         /**

--- a/tests/lib/rules/strict.js
+++ b/tests/lib/rules/strict.js
@@ -10,7 +10,8 @@
 //------------------------------------------------------------------------------
 
 const rule = require("../../../lib/rules/strict"),
-    { RuleTester } = require("../../../lib/rule-tester");
+    { RuleTester } = require("../../../lib/rule-tester"),
+    FlatRuleTester = require("../../../lib/rule-tester/flat-rule-tester");
 
 //------------------------------------------------------------------------------
 // Tests
@@ -655,6 +656,49 @@ ruleTester.run("strict", rule, {
             errors: [
                 { messageId: "unnecessaryInClasses", line: 2 },
                 { messageId: "multiple", line: 3 }
+            ]
+        }
+    ]
+});
+
+const flatRuleTester = new FlatRuleTester();
+
+// TODO: merge these tests into `ruleTester.run` once we switch to FlatRuleTester (when FlatRuleTester becomes RuleTester).
+flatRuleTester.run("strict", rule, {
+    valid: [
+        {
+            code: "'use strict'; module.exports = function identity (value) { return value; }",
+            languageOptions: {
+                sourceType: "commonjs"
+            }
+        },
+        {
+            code: "'use strict'; module.exports = function identity (value) { return value; }",
+            options: ["safe"],
+            languageOptions: {
+                sourceType: "commonjs"
+            }
+        }
+    ],
+
+    invalid: [
+        {
+            code: "module.exports = function identity (value) { return value; }",
+            options: ["safe"],
+            languageOptions: {
+                sourceType: "commonjs"
+            },
+            errors: [
+                { messageId: "global", line: 1 }
+            ]
+        },
+        {
+            code: "module.exports = function identity (value) { return value; }",
+            languageOptions: {
+                sourceType: "commonjs"
+            },
+            errors: [
+                { messageId: "global", line: 1 }
             ]
         }
     ]

